core/lb.go

```diff
@@ -0,0 +1,202 @@
+package core
+
+import (
+	"errors"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/lpms/ffmpeg"
+)
+
+var ErrTranscoderBusy = errors.New("TranscoderBusy")
+
+type LoadBalancedTranscoder interface {
+	Transcoder
+	Stop()
+}
+
+type newTranscoderFn func(device string, workDir string) LoadBalancedTranscoder
+
+type LoadBalancingTranscoder struct {
+	transcoders []string
+	workDir     string
+	newT        newTranscoderFn
+
+	// The following fields need to be protected by the mutex `mu`
+	mu       *sync.RWMutex
+	load     map[string]int
+	sessions map[string]*transcoderSession
+	idx      int // Ensures a non-tapered work distribution
+}
+
+func NewLoadBalancingTranscoder(devices string, workDir string, newTranscoderFn newTranscoderFn) Transcoder {
+	d := strings.Split(devices, ",")
+	return &LoadBalancingTranscoder{
+		transcoders: d,
+		workDir:     workDir,
+		newT:        newTranscoderFn,
+		mu:          &sync.RWMutex{},
+		load:        make(map[string]int),
+		sessions:    make(map[string]*transcoderSession),
+	}
+}
+
+func (lb *LoadBalancingTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+
+	lb.mu.RLock()
+	session, exists := lb.sessions[job]
+	lb.mu.RUnlock()
+	if exists {
+		glog.V(common.DEBUG).Info("LB: Using existing transcode session for ", session.key)
+	} else {
+		var err error
+		session, err = lb.createSession(job, fname, profiles)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return session.Transcode(job, fname, profiles)
+}
+
+func (lb *LoadBalancingTranscoder) createSession(job string, fname string, profiles []ffmpeg.VideoProfile) (*transcoderSession, error) {
+
+	lb.mu.Lock()
```

Comment:
if this lock is only protecting the creation of the session, should we unlock after line 82?

Reply:
Do you mean after assigning `lb.idx` on [line 84](https://github.com/livepeer/go-livepeer/pull/1124/files/4f28d093eb2051d2e45b6d4592b0177ff6d7724d#diff-da447e938cf8298a688acc7c4626255aR84)? (That needs to be protected too).

We could explicitly unlock, but after this point the function only starts a goroutine and exits. Feels more future-proof to `defer` the unlock.

Is there a case that could lead to unnecessary lock contention?